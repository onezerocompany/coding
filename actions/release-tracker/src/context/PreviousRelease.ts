/**
 * @file Contains the previous release definition.
 * @copyright 2022 OneZero Company
 * @license MIT
 * @author Luca Silverentand <luca@onezero.company>
 */

import type { Version } from '@onezerocompany/commit';

/** Definition of a previous release. */
export interface PreviousRelease {
  /** Hash of the release. */
  sha?: string | undefined;
  /** Version number of the release. */
  version?: Version | undefined;
}
