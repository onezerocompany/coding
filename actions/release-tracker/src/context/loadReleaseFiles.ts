/**
 * @file Contains the function to get the paths of all the release files.
 * @copyright 2022 OneZero Company
 * @license MIT
 * @author Luca Silverentand <luca@onezero.company>
 */

import { setFailed } from '@actions/core';
import { sync as glob } from 'glob';
import { getOptionalInput } from '../utils/getOptionalInput';

/** Definition of a release file. */
export interface ReleaseFile {
  /** Whether the file is attached to the release. */
  attached: boolean;
  /** The path of the file. */
  path: string;
}

/**
 * Function that reads the optional input "files" (new-line delimited) and resolves the glob patterns.
 * If no input is provided, it returns an empty array. If the input is provided,
 * it returns an array of paths and throws an error when the glob pattern does not
 * match any files or a file is not found. The.
 *
 * @returns The paths of the release files.
 * @example getReleaseFiles();
 */
export function loadReleaseFiles(): Array<{ attached: boolean; path: string }> {
  const files = getOptionalInput('files');
  if (files === null) return [];
  return files
    .split(/\r?\n/u)
    .map((pattern) => pattern.trim())
    .filter((pattern) => pattern)
    .reduce<string[]>((acc, pattern) => {
      const resolvedFiles = glob(pattern);
      if (resolvedFiles.length === 0) {
        setFailed(`No files found for glob pattern: ${pattern}`);
        process.exit(1);
      }
      return [...acc, ...resolvedFiles];
    }, [])
    .map((path) => ({ attached: false, path }));
}
