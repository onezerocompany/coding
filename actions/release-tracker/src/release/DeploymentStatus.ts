/**
 * @file Defines a deployment status.
 * @copyright 2022 OneZero Company
 * @license MIT
 * @author Luca Silverentand <luca@onezero.company>
 */

/** Status of the deployment. */
export enum DeploymentStatus {
  /** The deployment is pending. */
  pending = 'pending',
  /** The deployment is in progress. */
  inProgress = 'in_progress',
  /** The deployment was successful. */
  success = 'success',
  /** The deployment failed. */
  failure = 'failure',
  /** The deployment was cancelled. */
  cancelled = 'cancelled',
}
