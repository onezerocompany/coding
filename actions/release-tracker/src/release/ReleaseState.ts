/**
 * @file Defines the Release class.
 * @copyright 2022 OneZero Company
 * @license MIT
 * @author Luca Silverentand <luca@onezero.company>
 */

import {
  listCommits,
  ChangeLog,
  ChangeLogType,
  Version,
  getBumpForCommitList,
} from '@onezerocompany/commit';
import { debug, info } from '@actions/core';
import { getLastestRelease } from '../utils/octokit/getLastestRelease';
import { createRelease } from '../utils/octokit/createRelease';
import { isDefined } from '../utils/isDefined';
import { ReleaseAction } from './ReleaseAction';
import type { ReleaseEnvironment } from './ReleaseEnvironment';
import { parseReleaseEnvironmentsArray } from './ReleaseEnvironment';

/** Represents a release and its associated issue. */
export class ReleaseState {
  /** Release environments to deploy to. */
  public environments: ReleaseEnvironment[] = [];
  /** ID of the release object. */
  public releaseId?: number;
  /** ID of the related issue tracker. */
  public issueTrackerId?: string;
  /** Version number of release. */
  public version?: Version;

  /**
   * Determines the next action to take for this release.
   *
   * @returns The next action to take.
   * @example await determineAction();
   */
  public get nextAction(): ReleaseAction {
    if (!isDefined(this.releaseId)) return ReleaseAction.createRelease;
    return ReleaseAction.none;
  }

  /**
   * Converts a JSON string to a Release State.
   *
   * @param json - The JSON string to convert.
   * @returns The release state.
   * @example const release = ReleaseState.fromJson(json);
   */
  public static fromJson(json: string): ReleaseState | null {
    try {
      const release = new ReleaseState();
      const parsed = JSON.parse(json) as Record<string, unknown>;
      const { releaseId, issueTrackerId, environments } = parsed;
      if (typeof releaseId === 'number') release.releaseId = releaseId;
      if (typeof issueTrackerId === 'string')
        release.issueTrackerId = issueTrackerId;
      if (Array.isArray(environments))
        release.environments = parseReleaseEnvironmentsArray(environments);
      return release;
    } catch {
      return null;
    }
  }

  /**
   * Executes the next action for this release.
   *
   * @example await executeNextAction();
   */
  public async executeNextAction(): Promise<void> {
    debug(`Running next action... ${this.nextAction}`);
    switch (this.nextAction) {
      /** Trigger a release creation. */
      case ReleaseAction.createRelease:
        await this.createRelease();
        break;
      /** Handle any other cases. */
      default:
        break;
    }
    if (this.nextAction !== ReleaseAction.none) {
      await this.executeNextAction();
    }
  }

  /**
   * Creates a release for the current action.
   *
   * @example await this.createRelease()
   */
  private async createRelease(): Promise<void> {
    info('Creating release...');
    // Fetch the latest release from GitHub.
    const previousRelease = await getLastestRelease();
    const previousVersion = Version.fromString(
      previousRelease?.tag_name ?? '0.0.0',
    );
    info(`Previous release: ${previousVersion.displayString}`);

    // Fetch commits since last release ref.
    const commits = listCommits({
      beginHash: previousRelease?.tag_name,
    });
    info(`Found ${commits.length} commits since last release.`);
    debug(`Commits: ${JSON.stringify(commits)}`);

    // Generate changelog for release based on commits.
    const changelog = new ChangeLog({
      commits,
      type: ChangeLogType.internal,
      markdown: true,
    }).text;
    debug(`Changelog: ${changelog}`);

    // Determine the next version.
    const bump = getBumpForCommitList(commits);
    this.version = previousVersion.bump(bump);
    info(`Next version: ${this.version.displayString}`);

    // Create release.
    await createRelease({
      releaseState: this,
      changelog,
    });

    /*
     * Generate changelog (internal GitHub changelog).
     * Create release.
     * Attach `manifest.json` to release.
     */
  }
}
