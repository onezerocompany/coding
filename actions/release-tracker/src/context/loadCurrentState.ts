/**
 * @file Load the current release from an issue.
 * @copyright 2022 OneZero Company
 * @license MIT
 * @author Luca Silverentand <luca@onezero.company>
 */

import { context as githubContext } from '@actions/github';
import type { IssueCommentEvent } from '@octokit/webhooks-definitions/schema';
import type { ProjectManifest } from '@onezerocompany/project-manager';
import { ReleaseEnvironment } from '../release/ReleaseEnvironment';
import { ReleaseState } from '../release/ReleaseState';
import { updateStateFromComment } from '../release/updateStateFromComment';
import type { Context } from './Context';

/**
 * Creates a new release state.
 *
 * @param parameters - The parameters.
 * @param parameters.manifest - The project manifest.
 * @returns The release state.
 * @example const state = createReleaseState({ manifest });
 */
function newReleaseState({
  manifest,
}: {
  manifest: ProjectManifest;
}): ReleaseState {
  const newState = new ReleaseState();
  newState.environments = manifest.environments.map(
    (environment) =>
      new ReleaseEnvironment({
        id: environment.id,
        needs: environment.needs,
        githubName: environment.github_name,
        deployed: false,
        type: environment.type,
        changelog: {
          generate: environment.changelog.generate,
          headers: environment.changelog.headers,
          footers: environment.changelog.footers,
        },
      }),
  );
  return newState;
}

/**
 * Load the current release from an issue.
 *
 * @param parameters - The parameters for the function.
 * @param parameters.manifest - The manifest of the project.
 * @param parameters.previousState - The previous state of the release.
 * @param parameters.context - The GitHub context.
 * @returns The current release.
 * @example const release = loadCurrentRelease();
 */
export function loadCurrentState({
  manifest,
  previousState,
}: {
  manifest: ProjectManifest;
  previousState: ReleaseState | null;
  context: Context;
}): {
  state: ReleaseState;
  currentCommentText: string | null;
  currentCommentId: number | null;
} {
  let currentCommentText: string | null = null;
  let currentCommentId: number | null = null;
  const state =
    previousState === null ? null : ReleaseState.fromJson(previousState.json);
  if (state === null) {
    return {
      state: newReleaseState({ manifest }),
      currentCommentText,
      currentCommentId,
    };
  }

  if (githubContext.eventName === 'issue_comment') {
    const event = githubContext.payload as IssueCommentEvent;
    const didUpdate =
      event.action === 'edited' &&
      event.comment.body !== event.changes.body?.from;
    if (didUpdate) {
      // Update the state based on the comment
      currentCommentText = event.comment.body;
      currentCommentId = event.comment.id;
      updateStateFromComment({
        state,
        comment: event.comment.body,
        manifest,
        username: event.sender.login,
      });
    }
  }

  return { state, currentCommentText, currentCommentId };
}
