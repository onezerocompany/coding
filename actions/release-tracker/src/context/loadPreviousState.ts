/**
 * @file Function to load the previous state of the release.
 * @copyright 2022 OneZero Company
 * @license MIT
 * @author Luca Silverentand <luca@onezero.company>
 */

import { context } from '@actions/github';
import type {
  IssueCommentEvent,
  IssuesEvent,
} from '@octokit/webhooks-definitions/schema';
import { ReleaseState } from '../release/ReleaseState';
import { getContentBetweenTags } from '../utils/getContentBetweenTags';

/**
 * Loads the previous state of the release from the GitHub context.
 *
 * @returns The previous state of the release.
 * @example const previousState = loadPreviousState();
 */
export function loadPreviousState(): {
  state: ReleaseState | null;
  currentIssueText: string;
} {
  let content = '';
  if (context.eventName === 'issues') {
    content = (context.payload as IssuesEvent).issue.body;
  }
  if (context.eventName === 'issue_comment') {
    content = (context.payload as IssueCommentEvent).issue.body;
  }

  if (content.includes('<!-- JSON BEGIN') && content.includes('JSON END -->')) {
    const json = getContentBetweenTags(
      '<!-- JSON BEGIN',
      'JSON END -->',
      content,
    );
    return {
      state: ReleaseState.fromJson(json),
      currentIssueText: content,
    };
  }
  return {
    state: null,
    currentIssueText: content,
  };
}
