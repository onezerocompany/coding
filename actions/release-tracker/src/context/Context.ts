/**
 * @file Contains the context for the release tracker action.
 * @copyright 2022 OneZero Company
 * @license MIT
 * @author Luca Silverentand <luca@onezero.company>
 */

import { loadManifestFromProject } from '@onezerocompany/project-manager';
import type { ReleaseState } from '../release/ReleaseState';
import { loadCurrentState } from './loadCurrentState';
import { loadPreviousState } from './loadPreviousState';

/** Context for the action. */
export class Context {
  /** Static version of the context. */
  public static default = new Context();

  /** Manifest of the project with all the settings. */
  public static readonly projectManifest = loadManifestFromProject();

  /** Previous state of the release to compare to. */
  public previousState: ReleaseState | null = null;
  /** Current state of the release. */
  public curentState: ReleaseState | null = null;

  /**
   * Initializes the context for the action.
   *
   * @example await Context.default.initialize();
   */
  public async initialize(): Promise<void> {
    this.previousState = loadPreviousState();
    this.curentState = loadCurrentState({
      manifest: Context.projectManifest,
    });
  }
}
