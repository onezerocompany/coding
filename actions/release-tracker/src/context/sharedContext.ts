/**
 * @file Contains the shared context for the release tracker action.
 * @copyright 2022 OneZero Company
 * @license MIT
 * @author Luca Silverentand <luca@onezero.company>
 */

import { Context } from './Context';

export const sharedContext = new Context();
