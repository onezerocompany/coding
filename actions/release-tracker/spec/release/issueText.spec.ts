import { EnvironmentType } from '@onezerocompany/project-manager';
import { Commit } from '@onezerocompany/commit';
import { DeploymentStatus } from '../../src/release/DeploymentStatus';
import { ReleaseState } from '../../src/release/ReleaseState';
import { Context } from '../../src/context/Context';

describe('release issue text', () => {
  it('should generate issue text', () => {
    const state = new ReleaseState();
    state.environments = [
      {
        type: EnvironmentType.firebaseHosting,
        deployed: false,
        status: DeploymentStatus.pending,
        github_name: 'Firebase',
      },
    ];
    state.commits = [
      new Commit({
        hash: '1234567890',
        message: '⭐️ feat/new(project) new feature',
      }),
    ];
    expect(
      state.issueText({
        manifest: Context.projectManifest,
      }),
    ).toMatchSnapshot();
  });
});
