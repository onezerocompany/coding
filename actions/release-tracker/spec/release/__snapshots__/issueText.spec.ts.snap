// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`release issue text should generate issue text 1`] = `
"## Release Details

###### Edit the changelogs below and tick the checkboxes to release to each individual environment.

#### New Features
- project: [new feature](https://github.com/onezerocompany/coding/commit/1234567890)

---

<!-- section_begin:firebase -->
### Firebase Hosting

<!-- changelog_begin -->
\`\`\`
New Features:
 - new feature
\`\`\`
<!-- changelog_end -->
<!-- release_item_start -->
- [ ] Release to Firebase
<!-- release_item_end -->
<!-- section_end:firebase -->

---

Created from release: [0.0.1](https://github.com/onezerocompany/coding/releases/tag/0.0.1)"
`;
